### PR TITLE
Fix 3D move generator bug

### LINKED ISSUE #3821: Move generator assertions on 3D archs with different layouts per layer
From an FPL presentation, there are a couple of asserts that fire in the move generators when given an architecture with significantly different block maps on different layers. From the description it appears at least some move generators assert when they can't find a grid location of the appropriate type within the target region. That is wrong; they should return a failure code (which the move generators can already do).

#### Expected Behaviour
This should work.

#### Current Behaviour
An assert fires.

#### Possible Solution
Change the asserts to return a failure code like move abort. This should be easy, as I believe all the move generators already return a code.

#### Steps to Reproduce
I don't think we currently have a regtest checking a die stack with the same kinds of blocks on different layers, but significantly different arrangements. We should add a few tests like that:
 - same block types on different layers, but different arrangements
 - different types of blocks on different layers

That should reproduce this error, and those should be turned into regtests.
I'm also not sure how many > 2 layer tests we have; if none we should make some to check for small errors like this too, although that can be done for a separate issue.

#### Your Environment
This was the Lazagna fork of VTR/OpenFPGA, but that seems to be very close to the VTR master to this should reproduce.

### PR DESCRIPTION
**Pull request by AmirhosseinPoolad**
This PR hardens the 3D move generation code to not make certain assumption about the architecture. Previously some small part of the code base assumed that an (x, y) location would have the same block type on all layers. Fixes #3821.

### CODE REVIEW

File: vpr/src/place/move_utils.cpp [hunk 1]
```diff
@@ -1133,8 +1141,8 @@ std::vector<t_physical_tile_loc> get_compressed_loc(const t_compressed_block_gri
     const auto& compatible_layers = compressed_block_grid.get_layer_nums();
 
     for (const int layer_num : compatible_layers) {
-        // This would cause a problem if two blocks of the same types are on different x/y locations of different layers
-        compressed_locs[layer_num] = compressed_block_grid.grid_loc_to_compressed_loc({grid_loc.x, grid_loc.y, layer_num});
+        t_physical_tile_loc compressed_loc = compressed_block_grid.grid_loc_to_compressed_loc({grid_loc.x, grid_loc.y, layer_num});
+        compressed_locs[layer_num] = compressed_loc;
```
Comment (soheilshahrouz) [inline vpr/src/place/move_utils.cpp:1145]: What is the purpose of `compressed_loc`? Any reason you don't directly write the return value into `compressed_locs[layer_num]`?

File: libs/libarchfpga/src/physical_types.h [hunk 2]
```diff
@@ -780,10 +780,15 @@ struct t_physical_tile_loc {
         , layer_num(layer_num_val) {}
 
     // Returns true if this type location layer_num/x/y is not equal to ARCH_FPGA_UNDEFINED_VAL
-    operator bool() const {
+    inline bool is_valid() const {
         return !(x == ARCH_FPGA_UNDEFINED_VAL || y == ARCH_FPGA_UNDEFINED_VAL || layer_num == ARCH_FPGA_UNDEFINED_VAL);
     }
 
+    // Returns true if this type location layer_num/x/y is not equal to ARCH_FPGA_UNDEFINED_VAL
```
Comment (soheilshahrouz) [inline libs/libarchfpga/src/physical_types.h:787]: Could you replace this with a doxygen comment.

**Review by soheilshahrouz — COMMENTED** [review]
Thanks for fixing this bug.
I left two comments susggesting minor changes.

File: vpr/src/place/move_utils.cpp (hunk 1 above)
Comment (AmirhosseinPoolad) [reply vpr/src/place/move_utils.cpp:1145]: I initially added some logic/extra assertions here, but removed them later. This is left over. I'll revert it.

**Review by amin1377 — APPROVED** [review]
Thanks, Amir!

**Review by AlexandreSinger — APPROVED** [review]
LGTM, thanks @AmirhosseinPoolad
